File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/dependencytree/DependencyTree.java
package org.jetbrains.bsp.bazel.server.sync.dependencytree;

import io.vavr.Lazy;
import io.vavr.collection.HashSet;
import io.vavr.collection.Map;
import io.vavr.collection.Set;
import org.jetbrains.bsp.bazel.info.BspTargetInfo.Dependency;
import org.jetbrains.bsp.bazel.info.BspTargetInfo.TargetInfo;

public class DependencyTree {
  private final Set<String> rootTargets;
  private final Map<String, TargetInfo> idToTargetInfo;
  private final Map<String, Lazy<Set<TargetInfo>>> idToLazyTransitiveDependencies;

  public DependencyTree(Map<String, TargetInfo> idToTargetInfo, Set<String> rootTargets) {
    this.rootTargets = rootTargets;
    this.idToTargetInfo = idToTargetInfo;
    this.idToLazyTransitiveDependencies = createIdToLazyTransitiveDependenciesMap(idToTargetInfo);
  }

  private Map<String, Lazy<Set<TargetInfo>>> createIdToLazyTransitiveDependenciesMap(
      Map<String, TargetInfo> idToTargetInfo) {
    return idToTargetInfo.mapValues(this::calculateLazyTransitiveDependenciesForTarget);
  }

  private Lazy<Set<TargetInfo>> calculateLazyTransitiveDependenciesForTarget(
      TargetInfo targetInfo) {
    return Lazy.of(() -> calculateTransitiveDependenciesForTarget(targetInfo));
  }

  private Set<TargetInfo> calculateTransitiveDependenciesForTarget(TargetInfo targetInfo) {
    var dependencies = getDependencies(targetInfo);
    var strictlyTransitiveDependencies = calculateStrictlyTransitiveDependencies(dependencies);
    var directDependencies = calculateDirectDependencies(dependencies);

    return strictlyTransitiveDependencies.addAll(directDependencies);
  }

  private Set<TargetInfo> calculateStrictlyTransitiveDependencies(Set<String> dependencies) {
    return dependencies.flatMap(idToLazyTransitiveDependencies::get).flatMap(Lazy::get);
  }

  private Set<TargetInfo> calculateDirectDependencies(Set<String> dependencies) {
    return dependencies.flatMap(idToTargetInfo::get);
  }

  public Set<TargetInfo> transitiveDependenciesWithoutRootTargets(String targetId) {
    var target = idToTargetInfo.get(targetId);

    return target
        .toSet()
        .flatMap(this::getDependencies)
        .filter(this::isNotARootTarget)
        .flatMap(this::collectTransitiveDependenciesAndAddTarget);
  }

  private Set<String> getDependencies(TargetInfo target) {
    return HashSet.ofAll(target.getDependenciesList()).map(Dependency::getId);
  }

  private boolean isNotARootTarget(String targetId) {
    return !rootTargets.contains(targetId);
  }

  private Set<TargetInfo> collectTransitiveDependenciesAndAddTarget(String targetId) {
    var target = idToTargetInfo.get(targetId).toSet();
    var dependencies = idToLazyTransitiveDependencies.get(targetId).toSet().flatMap(Lazy::get);

    return dependencies.addAll(target);
  }
}

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/dependencytree/DependencyTree.kt
package org.jetbrains.bsp.bazel.server.sync.dependencytree

import org.jetbrains.bsp.bazel.info.BspTargetInfo.Dependency
import org.jetbrains.bsp.bazel.info.BspTargetInfo.TargetInfo

class DependencyTree(
        private val rootTargets: Set<String> = emptySet(),
        private val idToTargetInfo: Map<String, TargetInfo> = emptyMap(),
) {
    private val idToDirectDependenciesIds: Map<String, Set<String>>
    private val idToLazyTransitiveDependencies: Map<String, Lazy<Set<TargetInfo>>>

    init {
        idToDirectDependenciesIds = idToTargetInfo.entries.associate { (id, target) ->
            Pair(
                    id,
                    getDependencies(target)
            )
        }
        idToLazyTransitiveDependencies = createIdToLazyTransitiveDependenciesMap(idToTargetInfo)
    }

    private fun createIdToLazyTransitiveDependenciesMap(
            idToTargetInfo: Map<String, TargetInfo>
    ): Map<String, Lazy<Set<TargetInfo>>> =
            idToTargetInfo.mapValues { (_, targetInfo) ->
                calculateLazyTransitiveDependenciesForTarget(targetInfo)
            }

    private fun calculateLazyTransitiveDependenciesForTarget(
            targetInfo: TargetInfo
    ): Lazy<Set<TargetInfo>> =
            lazy { calculateTransitiveDependenciesForTarget(targetInfo) }

    private fun calculateTransitiveDependenciesForTarget(
            targetInfo: TargetInfo
    ): Set<TargetInfo> {
        val dependencies = getDependencies(targetInfo)
        val strictlyTransitiveDependencies = calculateStrictlyTransitiveDependencies(dependencies)
        val directDependencies = idsToTargetInfo(dependencies)
        return strictlyTransitiveDependencies + directDependencies
    }

    private fun calculateStrictlyTransitiveDependencies(
            dependencies: Set<String>
    ): Set<TargetInfo> =
            dependencies.flatMap {
                idToLazyTransitiveDependencies[it]?.value.orEmpty()
            }.toSet()

    private fun idsToTargetInfo(dependencies: Set<String>): Set<TargetInfo> =
            dependencies.mapNotNull(idToTargetInfo::get).toSet()

    private fun directDependenciesIds(targetIds: Set<String>) =
            targetIds.flatMap {
                idToDirectDependenciesIds[it].orEmpty()
            }.toSet()

    fun allTargetsAtDepth(depth: Int, targets: Set<String>): Set<TargetInfo> =
            if (depth < 0)
                idsToTargetInfo(targets) + calculateStrictlyTransitiveDependencies(targets)
            else
                allTargetsAtDepth(depth, setOf(), targets)

    private tailrec fun allTargetsAtDepth(depth: Int, searched: Set<TargetInfo>, targets: Set<String>): Set<TargetInfo> =
            if (depth == 0)
                searched + idsToTargetInfo(targets)
            else
                allTargetsAtDepth(depth - 1, searched + idsToTargetInfo(targets), directDependenciesIds(targets))

    fun transitiveDependenciesWithoutRootTargets(targetId: String): Set<TargetInfo> =
            idToTargetInfo[targetId]?.let(::getDependencies).orEmpty()
                    .filter(::isNotARootTarget)
                    .flatMap(::collectTransitiveDependenciesAndAddTarget).toSet()

    private fun getDependencies(target: TargetInfo): Set<String> =
            target.dependenciesList.map(Dependency::getId).toSet()

    private fun isNotARootTarget(targetId: String): Boolean =
            !rootTargets.contains(targetId)

    private fun collectTransitiveDependenciesAndAddTarget(targetId: String): Set<TargetInfo> {
        val target = idToTargetInfo[targetId]?.let(::setOf).orEmpty()
        val dependencies = idToLazyTransitiveDependencies[targetId]?.let(::setOf).orEmpty()
                .map(Lazy<Set<TargetInfo>>::value).flatten().toSet()
        return dependencies + target
    }
}
